design/m1_push_to_file_design.md

```diff
@@ -0,0 +1,1302 @@
+- [Backwards Compatibility](#backwards-compatibility)
+- [Affected Components](#affected-components)
+- [Test Plan](#test-plan)
+- [Logs](#logs)
+- [Documentation](#documentation)
+- [Version update](#version-update)
+- [Security](#security)
+- [Audit](#audit)
+- [Development Tasks](#development-tasks)
+- [Definition of Done](#definition-of-done)
+- [Solution Review](#solution-review)
+- [Appendix](#appendix)
+
+## Useful Links
+
+<table>
+<thead>
+<tr class="header">
+<th>Name</th>
+<th>Link</th>
+</tr>
+</thead>
+<tbody>
+<tr class="odd">
+<td>PRD: Conjur Developer Experience for K8s</td>
+<td><p><a href="https://cyberark365.sharepoint.com/:w:/r/sites/Conjur/_layouts/15/Doc.aspx?sourcedoc=%7B956DB935-7E3D-4D88-B964-188BCB6F7729%7D&file=Conjur%20Developer%20Experience%20for%20K8s%20PRD.docx&action=default&mobileredirect=true">link</a> (private)</p></td>
+</tr>
+<tr class="even">
+<td>Aha Card</td>
+<td><p><a href="https://cyberark.aha.io/epics/SCR-E-76">link</a> (private)</p>
+<p><em>Note: This design document covers work for "Milestone 1: Push to File" as defined in this Aha Card.</em></p></td>
+</tr>
+<tr class="odd">
+<td>Feature Doc</td>
+<td><a href="https://cyberark365.sharepoint.com/:w:/r/sites/Conjur/_layouts/15/Doc.aspx?sourcedoc=%7BB782E509-693F-4086-85A6-5D477A0F4ABD%7D&file=Feature%20Doc%20-%20Kubernetes%20Developer%20Experience%20M1.docx&action=default&mobileredirect=true&cid=048c1606-6533-443c-ba09-f91590c68095">link</a> (private)</td>
+</tr>
+<tr class="even">
+<td>Sample Manifests and Policies From Feature Spec</td>
+<td><a href="https://gist.github.com/alexkalish/8a514defa9e741800c095dded9837582">link</a></td>
+</tr>
+<tr class="odd">
+<td>Kubernetes Documentation:Annotations</td>
+<td><a href="https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/">link</a></td>
+</tr>
+<tr class="even">
+<td>Exposing Pod Annotations to Containers</td>
+<td><a href="https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/">link</a></td>
+</tr>
+<tr class="odd">
+<td>Secrets Provider Documentation</td>
+<td><a href="https://docs.cyberark.com/Product-Doc/OnlineHelp/AAM-DAP/11.2/en/Content/Integrations/Kubernetes_deployApplicationsConjur-k8s-Secrets.htm?tocpath=Integrations%7COpenShift%252C%20Kubernetes%252C%20and%20GKE%7CDeploy%20Applications%7C_____3">link</a></td>
+</tr>
+<tr class="even">
+<td>Secrets Provider application-mode Helm chart manifest</td>
+<td><a href="https://github.com/cyberark/secrets-provider-for-k8s/blob/master/helm/secrets-provider/templates/secrets-provider.yaml">link</a></td>
+</tr>
+</tbody>
+</table>
+
+## Background
+
+### Secrets Provider Push to File Support
+
+The goal of the Milestone 1 (M1) Push-to-File initiative is to enhance the
+Secrets Provider init container integration with an option for writing
+secrets values that have been retrieved from Conjur into files that can be
+directly accessed by an application that is running in the same Pod.
```

Comment:
In SP we also have the ability to deploy it as an application container so that it exists in the user's same namespace but isn't coupled with their actual application container as with the init container solution. How easy can Push to File be applied to the application container flow? Is this something we see value in?

Reply:
Unfortunately, this wouldn't be easy. The thing that makes Push to File viable when running the Secrets Provider as a container within the same Pod as the application is that they can share a Volume (SP can write a file, the application can directly read it).
Adding this functionality to the Secrets Provider running as a standalone application container would require a remote file-writing functionality, something akin to creating a Kubernetes CSI driver (see https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/). This would be possible, but one tricky part would be figuring how to allow the Secrets Provider (running as a CSI agent) to sort of proxy the machine identity of the application for which its retrieving from Conjur (to then write to a CSI volume).